ambry-frontend/src/main/java/com/github/ambry/frontend/PostDatasetsHandler.java

```diff
@@ -0,0 +1,203 @@
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+      this.uri = restRequest.getUri();
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      RestRequestMetrics requestMetrics =
+          frontendMetrics.postDatasetsMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), false);
+      restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+      RetainingAsyncWritableChannel channel = new RetainingAsyncWritableChannel(frontendConfig.maxJsonRequestSizeBytes);
+      restRequest.readInto(channel, fetchDatasetUpdateBodyCallback(channel));
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(frontendMetrics.postDatasetsSecurityProcessRequestMetrics,
+          securityCheckResult -> securityService.postProcessRequest(restRequest, securityPostProcessRequestCallback()),
+          uri, logger, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#postProcessRequest} finishes, add or update the dataset.
+     * @return a {@link Callback} to be used with {@link SecurityService#postProcessRequest}.
+     */
+    private Callback<Void> securityPostProcessRequestCallback() {
+      return buildCallback(frontendMetrics.postDatasetsSecurityPostProcessRequestMetrics, bytesRead -> {
+        ReadableStreamChannel outputChannel;
+        logger.debug("Got request for {} with payload", ACCOUNTS_CONTAINERS_DATASETS);
+        addOrUpdateDataset();
+        outputChannel = new ByteBufferReadableStreamChannel(ByteBuffer.allocate(0));
+        restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, RestUtils.JSON_CONTENT_TYPE);
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, outputChannel.getSize());
+        finalCallback.onCompletion(outputChannel, null);
+      }, uri, logger, finalCallback);
+    }
+
+    /**
+     * After reading the body of the account update request, inject the account and container in request for access check.
+     * @param channel the {@link RetainingAsyncWritableChannel} to read data out of.
+     * @return a {@link Callback} to be used with {@link RestRequest#readInto}.
+     */
+    private Callback<Long> fetchDatasetUpdateBodyCallback(RetainingAsyncWritableChannel channel) {
```

Comment:
nit: let's move this method right below start method so the order of methods follow the order of execution.

Reply:
Sure, updated.